bigtable-hbase-1.x-parent/bigtable-hbase-1.x-mapreduce/src/main/java/com/google/cloud/bigtable/mapreduce/validation/SyncTableValidationOnlyJob.java

```diff
@@ -0,0 +1,1102 @@
+    private ImmutableBytesWritable currentSourceHash;
+    private ImmutableBytesWritable nextSourceKey;
+    private BigtableTableHashAccessor.BigtableResultHasher targetHasher;
+
+    private Throwable mapperException;
+
+    public enum Counter {
+      BATCHES,
+      HASHES_MATCHED,
+      HASHES_NOT_MATCHED,
+      SOURCE_MISSING_ROWS,
+      SOURCE_MISSING_CELLS,
+      TARGET_MISSING_ROWS,
+      TARGET_MISSING_CELLS,
+      ROWS_WITH_DIFFS,
+      DIFFERENT_CELL_VALUES,
+      MATCHING_ROWS,
+      MATCHING_CELLS,
+      EMPTY_BATCHES,
+      RANGES_MATCHED,
+      RANGES_NOT_MATCHED
+    };
+
+    private static final NullWritable outputKey = NullWritable.get();
+    private Text outputValue = new Text();
+
+    @Override
+    protected void setup(Context context) throws IOException {
+
+      Configuration conf = context.getConfiguration();
+      sourceHashDir = new Path(conf.get(SOURCE_HASH_DIR_CONF_KEY));
+
+      // create target connection config
+      // inherit base config, but override connection based on job args
+      Configuration targetConf = new Configuration(conf);
+      targetConf.unset(ClusterConnection.HBASE_CLIENT_CONNECTION_IMPL);
+      if (!conf.onlyKeyExists(TARGET_BT_PROJECTID_CONF_KEY)) {
+        BigtableConfiguration.configure(
+            targetConf,
+            targetConf.get(TARGET_BT_PROJECTID_CONF_KEY),
+            targetConf.get(TARGET_BT_INSTANCE_CONF_KEY),
+            targetConf.get(TARGET_BT_APP_PROFILE_CONF_KEY, ""));
+      }
+
+      // create source connection config
+      // inherit base config, but override connection based on job args
+      Configuration sourceConf = new Configuration(conf);
+      sourceConf.unset(ClusterConnection.HBASE_CLIENT_CONNECTION_IMPL);
+      if (!conf.onlyKeyExists(SOURCE_BT_PROJECTID_CONF_KEY)) {
+        BigtableConfiguration.configure(
+            sourceConf,
+            sourceConf.get(SOURCE_BT_PROJECTID_CONF_KEY),
+            sourceConf.get(SOURCE_BT_INSTANCE_CONF_KEY),
+            sourceConf.get(SOURCE_BT_APP_PROFILE_CONF_KEY, ""));
+      }
+
+      // create target connection using target config
+      targetConnection = openConnection(targetConf, TARGET_ZK_CLUSTER_CONF_KEY);
+      targetTable = openTable(targetConnection, conf, TARGET_TABLE_CONF_KEY);
+
+      // create source connection using source config. if sourceConnection cannot be established,
+      // cell-level comparison is skipped and only hash comparison is performed
+      sourceConnection = null;
+      sourceTable = null;
+      try {
+        sourceConnection = openConnection(sourceConf, SOURCE_ZK_CLUSTER_CONF_KEY);
+        sourceTable = openTable(sourceConnection, conf, SOURCE_TABLE_CONF_KEY);
+
+        // verify source connection/table
+        Admin sourceAdmin = sourceConnection.getAdmin();
+        sourceAdmin.tableExists(sourceTable.getName());
+      } catch (Exception e) {
+        // explicitly set source as null and log source unavailable
```

Comment:
Maybe better to comment on why we need to set source as null, "set source to null" is obvious from the code.